### PR TITLE
Add key combinations for ctrl-delete and shift-delete

### PR DESCRIPTION
**Pull request by Syphdias**
Currently there is not option to bind ctrl-delete and shift-delete. As suggested by issue #3240, shift-delete could be used to bind "delete entry from history" as it is a common way to do so in other applications, e.g. browsers.

This, however, does only implement to use the key combination itself and does not assign a default action to any of them. This does enable to call one's all predefined actions. With the exec action this can expanded like the issue #3240 suggested.
If desirable, the key combinations could later get a default behavior.

I mainly wanted to use shift-delete like suggested in #3240. My approach would be to use the exec() action to build this functionality myself. I would not be comfortable deleting from the uses history due to expansion/escape issue and deleting too much or the wrong thing (for me I would risk it).

Two issues I would like to point to:
1. I am not sure which tests need to be expanded on and how. I could use some feedback on this, if the current tests, are not sufficient.
2. I modified the switch case for `Del`. In my testing, pressing only `Del` yielded an `r.buffer` with a length of 4. When combined with <kbd>Ctrl</kbd> or <kbd>Shift</kbd> the lenght went up to 6. Index 3 was always `;` though. So I skipped it for now. Should I check for for this explicitly and if so. Should I start a new switch to keep it analogous to other code and easier expansion, if there is ever another key combination that needs it?

### CODE REVIEW

**Comment by Syphdias** [comment]
Hey @junegunn, I was wondering if you find this PR useful.

Since https://github.com/junegunn/fzf/issues/1933 I assume the reason these key combinations are not already a thing, might be an issue with some terminals. Unfortunately, I cannot test if these keybinds work for iTerm or Terminal.app. 

Also, is there an option to use a "raw" mode for arbitrary keys? What I mean by that is to not be bound by the predefined keys manual definition, but use the escape sequence directly, e.g. `^[[3~` for <kbd>Delete</kbd>. I don't think this is possible right now. Issues I foresee would be that is is quite advanced and might lead to bug reports, reporting that some terminals behave different. This would need to be clarified in the man page.

TL;DR: Is this PR ([ctrl-delete and shift-delete](https://github.com/junegunn/fzf/pull/3284#top)) interesting to you? Would you be interested in accepting a PR that implements binding arbitrary keys?

**Comment by junegunn** [comment]
> Is this PR ([ctrl-delete and shift-delete](https://github.com/junegunn/fzf/pull/3284#top)) interesting to you?

I mainly use iTerm2 and it doesn't appear to recognize those keys, so I'm not personally going to use it. But I have confirmed that it works on Terminal.app, so I'm going to merge this.

**Comment by junegunn** [comment]
> Would you be interested in accepting a PR that implements binding arbitrary keys?

I honestly think it's a bit too much for this program.

**Comment by Syphdias** [comment]
> > Would you be interested in accepting a PR that implements binding arbitrary keys?
> 
> I honestly think it's a bit too much for this program.

Okay thanks, I'd rather know before I spend the time :)


> I mainly use iTerm2 and it doesn't appear to recognize those keys, so I'm not personally going to use it. But I have confirmed that it works on Terminal.app, so I'm going to merge this.

Should I add an disclaimer that some terminals might not recognise these key combinations or is this a general concern with all of the keys?

**Comment by junegunn** [comment]
Need to address these errors in tcell build (tcell is only used for Windows). Let me see...

```
jg@:~/github/fzf (add-ctrl-delete-and-shift-delete)> go run -tags tcell main.go
# github.com/junegunn/fzf/src/tui
src/tui/tcell.go:349:14: undefined: tcell.KeyCtrlDelete
src/tui/tcell.go:411:14: undefined: tcell.KeySDelete
```

**Comment by junegunn** [comment]
> Should I add an disclaimer that some terminals might not recognise these key combinations or is this a general concern with all of the keys?

The latter. Let's just see if anyone complains.

**Comment by junegunn** [comment]
Merged, thanks. Let me know if you have see issues.

**Comment by eggbean** [comment]
Is this supposed to work with the `ctrl-r` history? I would like to delete lines with typos, for example, but `shift-delete` doesn't do anything (v0.41.1). Do I have to the binding somehow? Thanks.

**Comment by eggbean** [comment]
Thanks, but I cannot get this to do what I thought it would.

    export FZF_CTRL_R_OPTS=" \
      --bind 'shift-delete:execute(echo {} >/tmp/test)'"

This writes the entry to `/tmp/test` but it still remains in the history list. I thought that the entry was supposed to be removed and I don't understand the need for the `/tmp/test` file.

**Comment by Syphdias** [comment]
This is not the solution to the problem of removing the entry from the history. And I don't have the solution for it.
I only provided a starting-off point, so you could tackle the problem yourself.

You need to find something to replace `echo {} >/tmp/test` with the right command/script, to do what you want.

**Comment by eggbean** [comment]
I see. Well I tried this hoping that it'd work (after backing up my history file), but it doesn't. I would prefer to avoid having to use a separate script.

    export FZF_CTRL_R_OPTS=" \
      --bind 'shift-delete:execute(LC_ALL=C sed -i "/{}/d" $HISTFILE)'"

ref: https://stackoverflow.com/a/13661794/140872

**Comment by eggbean** [comment]
It's possible that the command number changes before the command is executed, so I tried this:

    export FZF_CTRL_R_OPTS=" \
      --bind 'shift-delete:execute(str="$(echo {} | cut -d " " -f2-)" && LC_ALL=C sed -i "/$str/d" $HISTFILE)'"

Hopefully someone can come up with some better ideas.

**Comment by Syphdias** [comment]
If it was an easy problem, I would have included it in my PR :wink:.

What I can tell you is, that you have better chances with `'shift-delete:execute(LC_ALL=C sed -i /{}/d $HISTFILE)'` (no double quotes). But this is not perfect. It will fail for commands with slashes, quotes, variables (e.g. $1), and multi-line commands, which is not acceptable for me.

**Comment by eggbean** [comment]
I've made some progress (zsh):

```
export FZF_CTRL_R_OPTS=" \
  --bind 'shift-delete:execute(echo {} >/tmp/test)'"

% cat fzf-history-delete.sh
#!/bin/bash -x
str="$(echo "$*" | awk '{$1=""}1' | colrm 1 1 | sed -e 's/[]\/$*.^[]/\\&/g')"
LC_ALL=C sed -i "/$str$/d" "$HISTFILE"

% cp $HISTFILE{,.bak}

* Use `Ctrl-R` , `shift-delete`

% cat /tmp/test
 1774  fc -p $HISTFILE

% HISTFILE=$HISTFILE ./fzf-history-delete.sh $(cat /tmp/test)
++ echo '1774 fc -p $HISTFILE'
++ colrm 1 1
++ awk '{$1=""}1'
++ sed -e 's/[]\/$*.^[]/\\&/g'
+ str='fc -p \$HISTFILE'
+ LC_ALL=C
+ sed -i '/fc -p \$HISTFILE$/d' /home/jason/.cache/zsh/zsh_history

% fc -R $HISTFILE
```

This seems to work with everything I've tried with it, in zsh. Doesn't seem to work in bash.

```
export FZF_CTRL_R_OPTS=" \
  --bind 'shift-delete:execute(HISTFILE=$HISTFILE fzf-history-delete.sh {})'"
```

This doesn't seem to work.

I tried making it script-less, but this nuked the history file.

```
export FZF_CTRL_R_OPTS=" \                                                                                   
  --bind 'shift-delete:execute(HISTFILE=$HISTFILE str="$(echo {} | awk '{$1=""}1' | colrm 1 1 | sed -e 's/[]\/$*.^[]/\\&/g')" ;\
    LC_ALL=C sed -i "/$str$/d" "$HISTFILE")'"
```

I'll add a test to avoid empty `$str` and simplify the `echo|awk|colrm` with `shift`.

**Comment by eggbean** [comment]
I've made some progress (zsh):

```
export FZF_CTRL_R_OPTS=" \
  --bind 'shift-delete:execute(echo {} >/tmp/test)'"
```

```
% cat fzf-history-delete.sh
#!/bin/bash -x
str="$(echo "$*" | awk '{$1=""}1' | colrm 1 1 | sed -e 's/[]\/$*.^[]/\\&/g')"
LC_ALL=C sed -i "/$str$/d" "$HISTFILE"

% cp $HISTFILE{,.bak}

* Use `Ctrl-R` , `shift-delete`

% cat /tmp/test
 1774  fc -R $HISTFILE

% HISTFILE=$HISTFILE ./fzf-history-delete.sh $(cat /tmp/test)
++ echo '1774 fc -R $HISTFILE'
++ colrm 1 1
++ awk '{$1=""}1'
++ sed -e 's/[]\/$*.^[]/\\&/g'
+ str='fc -R \$HISTFILE'
+ LC_ALL=C
+ sed -i '/fc -p \$HISTFILE$/d' /home/jason/.cache/zsh/zsh_history

% fc -R $HISTFILE   # Reload history
```

This seems to work with everything I've tried with it, in zsh. Doesn't seem to work in bash.

```
export FZF_CTRL_R_OPTS=" \
  --bind 'shift-delete:execute(HISTFILE=$HISTFILE fzf-history-delete.sh {})'"
```

This doesn't seem to work.

I tried making it script-less, but this nuked the history file.

```
export FZF_CTRL_R_OPTS=" \
  --bind 'shift-delete:execute(str="$(echo {} | awk '{$1=""}1' | colrm 1 1 | sed -e 's/[]\/$*.^[]/\\&/g')" ; \
    LC_ALL=C sed -i "/$str$/d" "$HISTFILE"; \
    fc -R "$HISTFILE")'"
```

I'll add a test to avoid empty `$str` and simplify the `echo|awk|colrm` with `shift`.

**Comment by eggbean** [comment]
I've made some progress (zsh):

```
export FZF_CTRL_R_OPTS=" \
  --bind 'shift-delete:execute(echo {} >/tmp/test)'"
```

```
% cat fzf-history-delete.sh
#!/bin/bash -x
str="$(echo "$*" | awk '{$1=""}1' | colrm 1 1 | sed -e 's/[]\/$*.^[]/\\&/g')"
LC_ALL=C sed -i "/$str$/d" "$HISTFILE"

% cp $HISTFILE{,.bak}

* Use `Ctrl-R` , `shift-delete`

% cat /tmp/test
 1774  fc -R $HISTFILE

% HISTFILE=$HISTFILE ./fzf-history-delete.sh $(cat /tmp/test)
++ echo '1774 fc -R $HISTFILE'
++ colrm 1 1
++ awk '{$1=""}1'
++ sed -e 's/[]\/$*.^[]/\\&/g'
+ str='fc -R \$HISTFILE'
+ LC_ALL=C
+ sed -i '/fc -p \$HISTFILE$/d' /home/jason/.cache/zsh/zsh_history

% fc -R $HISTFILE   # Reload history
```

This seems to work with everything I've tried with it, in zsh. Doesn't seem to work in bash.

```
export FZF_CTRL_R_OPTS=" \
  --bind 'shift-delete:execute(HISTFILE=$HISTFILE fzf-history-delete.sh {})'"
```

This doesn't seem to work.

I tried making it script-less, but this nuked the history file.

```
export FZF_CTRL_R_OPTS=" \
  --bind 'shift-delete:execute(str="$(echo {} | awk '{$1=""}1' | colrm 1 1 | sed -e 's/[]\/$*.^[]/\\&/g')" ; \
    LC_ALL=C sed -i "/$str$/d" "$HISTFILE"; \
    fc -R "$HISTFILE")'"
```

I'll should test to avoid empty `$str`, or whatever is going on and simplify the `echo|awk|colrm` with `shift`.
